src/main/java/io/github/thebusybiscuit/slimefun4/core/services/JsonDeserializationService.java

```diff
@@ -0,0 +1,160 @@
+package io.github.thebusybiscuit.slimefun4.core.services;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonPrimitive;
+import me.mrCookieSlime.Slimefun.api.Slimefun;
+import org.bukkit.configuration.serialization.ConfigurationSerialization;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.logging.Level;
+
+/**
+ * Represents a mapper between classes and json deserialization logic. This class is
+ * somewhat similar to Bukkit's {@link ConfigurationSerialization} class.
+ *
+ * @author md5sha256
+ */
+public class JsonDeserializationService {
+
+    public static final String ENUM_CLASS_KEY = "enum-class", ENUM_VALUE_KEY = "value";
+
+    private final JsonParser parser = new JsonParser();
+    private final Map<String, Function<JsonElement, ?>> deserializerMap = new HashMap<>();
+
+    public JsonDeserializationService() {
+    }
```

Comment:
This constructor is redundant, an empty public constructor is automatically inferred anyway.
```suggestion
    public JsonDeserializationService() {}
```